front/lib/api/assistant/conversation.ts

```diff
@@ -993,32 +993,22 @@ export async function* editUserMessage(
     };
     return;
   }
-  if (message.mentions.filter((m) => isAgentMention(m)).length > 0) {
-    yield {
-      type: "user_message_error",
-      created: Date.now(),
-      error: {
-        code: "not_allowed",
-        message:
-          "Editing a message that already has agent mentions is not yet supported",
-      },
-    };
-    return;
-  }
 
+  // Only the last message of the conversation can be edited. In that case, a new version of the message will be created.
+  // Agent messages answering previous versions of the message will be kept but will not be displayed - only agent message
+  // having a visible parent will be shown.
+  const userMessages = conversation.content.filter((messages) =>
```

Comment:
I think this is not resilient to concurrent edits or concurrent post user messages right? It's good to have that error early but we need to enforce this within the transaction (that holds an advisory lock below).

If you look l.1083 we check that there isn't a new verison of that user message already which covers concurrent edit message requests.

But if you look 1144 we have a comment that relies on the invariant that there should be no agent messages after the currently edited user message which is not true here:
```
 // For now agent messages are appended at the end of conversation
// it is fine since for now editing with new mentions is only supported
// for the last user message
 ```
 
 As a consequence the computation of `nextMessageRank` is likely wrong here. We want instead to add a new version to the rank of the next agent message.
 
 Since you were hidding the messages without parent in the UI you didn't see the bug but this looks incorrect and require a bit of refactor + checking that the last message is an agent message and its rank is the edited user message's rank + 1. We also want to error if that's not the case from within the transaction to make sure we're resilient to concurrent edits and post user messages.

Reply:
As per IRL there is forward-proof (tree of conversations) with our current model that does not imply having holes in ranks when rendering conversations.

Goal is to explore allocating always new ranks here + fix conversation rendering + evaluate impact on pagination

Reply:
I moved the test into the transaction to better support concurrent edits. 

As discussed, the `newMessageRank` will be a new rank added at the end of the conversation, leaving the previous agent responses as is. Now the `getMaxRankMessages`, used in both ui and api, is filtering the messages that should be displayed. 

I've checked usages of the rank and it seems ok to have "holes" when getting list of displayed messages. Pagination is actually not much impacted, as we are using an infinite scroll and passing the last rank that was displayed - we're still getting the n next visible messages.